### PR TITLE
refactor(post_regression): ask parameters once and clarify pipeline

### PR DESCRIPTION
**Pull request by felipehuerta17**
**Refactor fit and post regression functions to facilitate API usage & code testing**

Following the comments of @sasalatart on #76 regarding pylint complaining for the number of arguments, I refactored model, .fit and .set_params to ask for
parameters once. The typical usage of the API, as all our examples
show, follows this pipeline:

1. Build an empty model
2. Set parameters and initial conditions
3. Fit the model
4. Calculate confidence intervals.

Currently, we were asking the user for (t,X) on steps 3 and 4, and for the population on 3,4 but it was implicitely calculated on 2.  This PR refactors the code in a way that the population is calculated directly from the IC on 2, and we don't ask the user again twice. Additionally, (t,X) is asked only on step 3. This enables to ask the user three less parameters on fit and the post_regression routines.

**Commit f3d93f4** 
Ask the user for the population only once when building the model and using the set_params / set_parameters functions. Therafter, when fitting the population parameters is
no longer required as it will be stored in self.pop

**Commit e9558d6** 
This commit introduces a new dict attribute for the model,
called self.postreg. This dict stores user inputs from the
first call of the fit function: n_obs and t_obs, in a similar
way that fit_index is stored.

The functions model and post_regression are updated and the syntax
is considerably simplified, which should help code maintenance and
API usage.

The test_model.py is updated to consider that the fit function, from
the previous commits, use less parameters.

The notebooks SIR and SIR-X are updated to be consistent with these
changes.

This is the last building block (It took a while!) before finishing #70 and finally
being ready to go for the release

### CODE REVIEW

File: doc/tutorials/SIR.ipynb [hunk 1]
```diff
@@ -507,14 +507,49 @@
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "tags": [
+     "outputPrepend",
```
Comment (sasalatart) [inline doc/tutorials/SIR.ipynb:512]: What is this for?

Comment (felipehuerta17) [reply doc/tutorials/SIR.ipynb]: This is again something that vscode adds and makes the docs crash.

File: opensir/models/model.py [hunk 2]
```diff
@@ -217,8 +220,12 @@ def fit(self, t_obs, n_obs, fit_index=None):
 
         # Check consistent inputs for fitting
         # for i in
-
+        # Set post-fit model attributes
         self.fit_index = fit_index
+        if self.postreg["t_obs"] is None:
+            self.postreg["t_obs"] = t_obs
```
Comment (jia200x) [inline opensir/models/model.py:226]: why are we storing `t_obs` and `n_obs`? In practice the user already have pointers to these arrays. What's the benefit of storing it in the model?

File: opensir/models/model.py [hunk 3]
```diff
@@ -39,12 +39,14 @@ def __init__(self):
...
         self.pop = None
         self.w0 = None
-        # Attributes setted after fitting
-        self.pcov = None
         self.fit_input = None
         self.fit_index = None  # Store fitting info for post-regression analysis
-        self.t_obs = None
-        self.n_obs = None
+        # Attributes setted after fitting
+        self.postreg = {
```
Comment (jia200x) [inline opensir/models/model.py:45]: See https://github.com/open-sir/open-sir/pull/77/commits/20387251101e11f3ab8ec55436bc7b90bad701fa#r418971623.
IMO the only one that makes sense to keep here is "pcov", which could be a simple member

Comment (jia200x) [reply opensir/models/model.py:45]: ok, I see the spirit behind these (they are used by `ci_xxx` functions).

However, in this case I think I would drop the dict and leave it as it was before (simple members). IMO this adds unnecessary maintaining overhead

Comment (felipehuerta17) [reply opensir/models/model.py:45]: pylint complains for too much attributs, that was my motivation of creating this new structure

Comment (jia200x) [reply opensir/models/model.py:45]: Adding a dict removes the pylint warning but the spirit behind pylint's complain is the same :) We are adding 3 new variables anyway.

I will remove `fit_input` and `fit_index` in a follow up (WIP), so pylint shouldn't complain here.

Comment (felipehuerta17) [reply opensir/models/model.py:45]: OK, so should I wait your follow up?

Where are you going to store fit_input and fit_index?

**Comment by jia200x** [comment]
ok, it makes sense now to put all "fit" stuff together, so in the future they can be represented with a "FitContainer" class.

File: opensir/models/model.py [hunk 4]
```diff
@@ -39,9 +39,14 @@ def __init__(self):
         self.p = None
         self.pop = None
         self.w0 = None
-        self.pcov = None
-        self.fit_input = None
-        self.fit_index = None  # Store fitting info for post-regression analysis
+        # Attributes setted after fitting
+        self.fit_attr = {
```
Comment (jia200x) [inline opensir/models/model.py:43]: I guess it's better here to define "self.fit_attr" to None. Then, the "fit" function creates the dictionary and assigns it directly.

File: opensir/models/model.py [hunk 5]
```diff
@@ -206,29 +212,33 @@ def fit(self, t_obs, n_obs, population, fit_index=None):
         if len(t_obs) != len(n_obs):
             raise self.InconsistentDimensionsError(Exception)
         # if no par_index is provided, fit only the first parameter
-        if (fit_index is None) & (self.fit_index is None):
+        if (fit_index is None) & (self.fit_attr["fit_index"] is None):
```
Comment (jia200x) [inline opensir/models/model.py:215]: note the "&" is a bitwise and, not a "logical" and. This "&" might produce unexpected results

Comment (jia200x) [reply opensir/models/model.py:215]: in fact, maybe it's better to drop the second clause. The "fit_index" attribute is set anyway

Comment (felipehuerta17) [reply opensir/models/model.py:215]: Good point, with the refactor that you suggest it is no longer necessary. This is updated.

File: opensir/models/model.py (hunk 4 above)
Comment (felipehuerta17) [reply opensir/models/model.py:43]: Alright. The fit function will create the dictionary. I noticed that there was one attribute that doesn't depends on the fit, but it depends on the model, and it is fit_input. I included it on the model attributes again.

**Comment by jia200x** [comment]
ok, I think all changes are addressed. Thanks!

**Pull request merged**